source/MaterialXCore/Document.cpp

```diff
@@ -1061,6 +1061,45 @@ void Document::upgradeVersion(bool applyFutureUpdates)
                     }
                 }
             }
+
+            // Make it so that interface names and nodes in a nodegraph are not duplicates
+            // If they are, rename the nodes.
```

Comment:
I wonder if it's enough to only check the used interfacenames? If the nodegraph has a nodedef with inputs that are never connected inside with interfacename... In theory the nodegraph could then still have inputs that are named the same as internal nodes. Not sure if this will ever be a problem. But to cover all cases, maybe the nodedef's input/parameter names should be added to the interfaceNames set below (if the nodegraph has a nodedef assigned that is).

Reply:
Maybe what we do here is (Jonathan + I discussion)

1) Detect+fix: if there is a nodedef input but no interface name, this is illegal and we just delete those inputs and the ones on the nodegraph as well. This is an "upgrade" to fix.
2) Validate-only: Otherwise we detect this in validation which occurs before "upgrade" in which case we just throw an error.

For overall logic, propose to pull this out as a new method on Document which can be called from here. If we want to migrate checking to validation then we can consider how to reuse the utility.

To check: the proposal would be to do 1) for now and not 2) though it's arguable that we are "fixing things" without the users explicit knowledge.

Reply:
Additionally the 3rd case is:

1) nodegraph inputs which are not connected to any internal node. (new situation)
2) nodegraph outputs which are not connected to any internal node (existing situation)

If we delete these in the nodegraph we also need to delete on any corresponding nodedef.
This can be handled in "upgrade" or when we "validate". Seems better to do in validation.

The 1st case is the current duplicate checking code.

Reply:
Yet another point to consider is for now avoid overcomplcating things there is a user case for having a "template" nodegraph which has all possible inputs/outputs/and interface names and it is reused by > 1 nodedef which choses a subset to "publish". But let's not go there for now :).

Reply:
@bernardkwok I wonder if deleting unused/unconnected ports from the nodedef is taking it too far.. I think having a nodedef with ports that are not used inside the nodegraph implementation is a perfectly valid setup. For example, if the nodegraph is an implementation for a specific target, let's say a low-end renderer with approximations, then an input like `subsurface_radius` may very well be unused/unconnected inside this graph. But we can't delete it form the nodedef because then we destroy all other uses of this nodedef.

So I think just renaming any conflicting internal nodes based on the union of nodedef inputs and all used interfacenames, is the right approach here. The only addition in code here is to add all nodedef ports to the StringSet used for finding conflicting nodes.

Reply:
Hey @niklasharrysson , I though that the nodedef inputs must match the interface names in the nodegraph otherwise it's a signature mismatch. If it can be a subset than that's fine. I still think that if we have a nodedef input which references an interface name not found then it's illegal.

Seems like we can get in this part first and discuss some more offline? :)

Reply:
@bernardkwok Sure, the code is good enough as is for now. And we can discuss the other case offline :)